### PR TITLE
builder filter volumes on run

### PR DESCRIPTION
**Pull request by cpuguy83**
Fixes #3639

This makes the builder, on `RUN` commands disable volumes, this way
anything written to dirs that have a volume will persist in the image,
as is the expected behavior in builds.

This is an alternative to #7133
This one is a little hacky but seems to work without modifying how
containers actually work in Docker (as #7133 does)
This is a last ditch effort to fix this within just the builder
itself.

### CODE REVIEW

File: integration-cli/docker_cli_build_test.go [hunk 1]
```diff
@@ -5115,3 +5115,32 @@ func TestBuildNotVerbose(t *testing.T) {
...
+		map[string]string{"baz": "qux"})
+	defer ctx.Close()
+
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	if _, err = buildImageFromContext("test", ctx, false); err != nil {
+		t.Fatalf("expected build to succeed:\n%v", err)
+	}
```
Comment (duglin) [inline integration-cli/docker_cli_build_test.go]: Should it test that the data is still around on subsequent uses of this image?

Comment (duglin) [inline integration-cli/docker_cli_build_test.go]: Ah never mind, I forgot that each RUN will be like a new container from the previous image

**Comment by duglin** [comment]
I want to make sure I grok this. Using this Dockerfile:
```
FROM ubuntu
RUN mkdir /foo
RUN echo foo > /foo/foo
VOLUME /foo
RUN cat /foo/foo
RUN echo hello world >> /foo/foo
RUN cat /foo/foo
```
shouldn't /foo/foo at then end contain "foohello world"?
I'm seeing just 'foo'

**Comment by crosbymichael** [comment]
seems sane to me

File: builder/dispatchers.go [hunk 2]
```diff
@@ -253,6 +253,16 @@ func run(b *Builder, args []string, attributes map[string]bool, original string)
...
 	defer c.Unmount()
 
+	// Unset the volumes
+	volumeConfig := c.Config.Volumes
+	c.Config.Volumes = nil
+	volumes := c.Volumes
+	c.Volumes = nil
+	defer func() {
+		c.Volumes = volumes
+		c.Config.Volumes = volumeConfig
```
Comment (crosbymichael) [inline builder/dispatchers.go]: Shouldn't this happen before the `b.commit`

Comment (cpuguy83) [inline builder/dispatchers.go]: oh poo, yep. I moved this from somewhere else and never updated.
Fixed now and beefed up the test which didn't catch this before.

**Comment by cpuguy83** [comment]
@duglin You are correct, there was an issue, and the test wasn't beefy enough.

File: builder/dispatchers.go [hunk 3]
```diff
@@ -253,10 +253,19 @@ func run(b *Builder, args []string, attributes map[string]bool, original string)
 	c.Mount()
 	defer c.Unmount()
 
+	// Unset the volumes
+	volumeConfig := c.Config.Volumes
+	c.Config.Volumes = nil
+	volumes := c.Volumes
+	c.Volumes = nil
+
 	err = b.run(c)
```
Comment (duglin) [inline builder/dispatchers.go]: if there's an error do we need to reset things via a defer? Or is just returning ok?

File: builder/internals.go [hunk 4]
```diff
@@ -90,6 +90,16 @@ func (b *Builder) commit(id string, autoCmd []string, comment string) error {
 			return err
 		}
 		defer container.Unmount()
+
+		// Make sure all resources are setup, but we can ignore errors
+		// Supress output from this
```
Comment (duglin) [inline builder/internals.go]: Just curious, why do we want to ignore errors here?

File: builder/dispatchers.go (hunk 3 above)
Comment (cpuguy83) [inline builder/dispatchers.go]: I was thinking about this... technically I think we'd want the real information about the container that actually ran.
wdyt?

File: builder/internals.go (hunk 4 above)
Comment (cpuguy83) [inline builder/internals.go]: Specifically I was seeing an error in the tests where:
```Dockerfile
FROM scratch
MAINTAINER jpetazzo@docker.com
```
Since `MAINTAINER` the maintainer line is trying to be run with `/bin/sh`, but `/bin/sh` doesn't exist in `scratch`

**Comment by duglin** [comment]
First, sorry if I get some of this wrong, I'm still trying to get my head around this issue/PR, but I'm not keen on this PR and in fact I'm not sure I even agree with the original issue in https://github.com/docker/docker/issues/3639 .

I believe what that issue is asking for is that a volume's data live on beyond the life-span of a container, when by definition its not supposed to. Each time I create a new container from an image that had a VOLUME in its Dockerfile I'm supposed to get a new volume at that mount point.  The Dockerfile build process does exactly that. It creates a new container from an image. The idea that the new container should have VOLUME data from a previous instance of that image seems to break how VOLUME is supposed to work.

While I understand there is the desire to have data in these volumes persist for more than one RUN command, I think it would be better to solve this w/o breaking some of the basic design principles around how Docker works today.  In other words, I don't think "docker build" should be able to create new containers (which is what it does on each RUN step) and have data appear that I wouldn't be able to reproduce via normal "docker" CLI commands.

To me, this:
```
FROM ubuntu
VOLUME /foo
RUN echo hi > /foo/foo
RUN cat /foo/foo
```
should fail on the last RUN because its a new container with no data (no /foo/foo) from the previous RUN.

If we really want to do this then let's come up with something to put in the Dockerfile that means "execute these commands within the same container" - I think that's a more consistent model to expose to the user.

**Comment by cpuguy83** [comment]
I think it's completely counter-intuitive (and even a waste of resources since the data from the image is copied to the volume) to create volumes during build at all.

**Comment by thaJeztah** [comment]
> I think it's completely counter-intuitive

I think that's the main issue here. Although @duglin is correct from a *technical viewpoint*; each step is a new container, so should *act* like a container; From an end-user's perspective that is just an implementation-detail.

I think most users (especially when *new* to Docker) expect an volume ***during build*** to be only a ***"declaration"***, i.e. the volume won't be applied until *I* start a container from the image, but during build it's just a regular directory *inside* the container.

In short I think most users will expect it to work like;

```
- build-step
- build-step
- ** declare ** volume
- build step
-...
- tag/commit image
```

This is especially important for users that build **from** an (official) image that uses a volume; there's no way to "undo" that and effectively, you cannot do anything with that directory inside your `Dockerfile`.

Being able to `NOVOLUME` (previously `UNSET`) the volume (https://github.com/docker/docker/pull/8177) would help, but also becomes messy if the volume has to be enabled again at the end. Perhaps https://github.com/docker/docker/pull/10775 could include support for a `--no-volume=/foo/bar` option to disable (a) volumes(s) on a per-step base

> I don't think "docker build" should be able to create new containers (which is what it does on each RUN step) and have data appear that I wouldn't be able to reproduce via normal "docker" CLI commands.

*That's* something I agree with. Could that be solved with a `--no-volumes` flag for `docker create` and `docker run`? Perhaps there should be `--enable-volumes` flag to `docker build` as well, for people that want the old behavior?

**Comment by duglin** [comment]
If you look at the original issue, and the small example in my previous comment, it appears that people are expecting the volume to be there immediately and the data to be saved in the image. If we change it to be just a declaration will those people be upset?  If not then I wonder why the issue was opened to begin with - the current behavior should be ok for them.  

While I can understand thinking of the build process as being a bit special, if you convert a Dockerfile that has a VOLUME and 3 RUNS into 4 images/Dockerfiles that are built one after the other, will it have the same results as if it were all in one Dockerfile/build?  Should it?  Right now I think it does.  And I like that consistency.  Do we want to open the door for more reasons to make them different?

Based on @cpuguy83's comment about how creating volumes during the build is a waste of time/resources, I wonder if the real issue here is one of education.  Meaning, should the solution be to modify the docs to say something like:
```
Volumes created during the build process via the VOLUME command, while immediately visible, should not be used during the build process. Any data written to them will be transient and not visible on subsequent Dockerfile commands. This is because data written to volumes are *NOT* saved in images and containers created from those images will be allocated new volumes each time.
```

**Comment by cpuguy83** [comment]
@duglin Yes, people expect `VOLUME` to really do nothing in the context of a build other than actually create a volume when someone creates a "real" container from the image.

The current way the builder works (ie, each command is a container -> commit to image) is strictly for caching purposes.  Because of this we have things like `ENV`, `EXPOSE`, and `VOLUME` that create an image layer needlessly with no actual filesystem changes. These are just configurations that get used in a `RUN` command... and I guess the argument here is that people don't want to use `VOLUME` in the context of a `RUN`.

**Comment by cpuguy83** [comment]
And I say `in the context of a run` because that's the only thing that actually triggers the volume.

**Comment by thaJeztah** [comment]
@duglin not sure I understand your comment here:

> it appears that people are expecting the volume to be there immediately and the data to be saved in the image.

I'm not sure I read the same in there; I just re-read the original ticket https://github.com/docker/docker/issues/3639 and afaict, people *don't* expect a volume to be created during build; they expect the data to be *mutable* and part of the image.

Especially this; https://github.com/docker/docker/issues/3639#issuecomment-47534519 the user is expecting the `/tmp/docker/hello` to be removed from the image, but it is still there.

> If we change it to be just a declaration will those people be upset?

No, I think that's what is being asked. Perhaps I wasn't clear on my intent what a `VOLUME` declaration would do;

As an end-user, I expect a `VOLUME` to do *nothing* during build, but when I start a container from the image, the volume should be created and the data copied into the volume.

Basically, "disable" creating/mounting volumes during build.

**Comment by duglin** [comment]
@cpuguy83 finally got thru to me (via irc) on what's going on and I now understand it better.  Sorry for the noise/misunderstanding.

**Comment by thaJeztah** [comment]
@duglin no problem! It can be difficult to 'grasp'. :dancer:

**Comment by cpuguy83** [comment]
Closing this one since it seems like a no-go.

**Comment by thaJeztah** [comment]
@cpuguy83 bummer, why?
